backends/s3/s3.go

```diff
@@ -109,35 +109,35 @@ func (b *Backend) List(ctx context.Context, prefix string) (simpleblob.BlobList,
 	}
 
 	// Request and cache full list, and use marker file to invalidate the cache
-	now := time.Now()
 	data, err := b.Load(ctx, UpdateMarkerFilename)
-	if err != nil && !os.IsNotExist(err) {
-		return nil, err
+	exists := true
+	if err != nil {
+		if !os.IsNotExist(err) {
+			return nil, err
+		}
+		exists = false
 	}
 	current := string(data)
 
 	b.mu.Lock()
-	lastMarker := b.lastMarker
-	age := now.Sub(b.lastTime)
+	shouldUpdate := b.lastList == nil || b.lastMarker != current ||
+		time.Since(b.lastTime) >= b.opt.UpdateMarkerForceListInterval ||
+		!exists
+	blobs := b.lastList
```

Comment:
Good change, this guarantees that we take a reference to the blobs at the time we checked the other state.

It probably would not have mattered that we previously could have returned a newer version if they got updated in a different goroutine, but this is cleaner.